File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/stored/library/items/files/job/GivenAFileThatDoesNotYetExist/AndTheInputStreamCannotBeOpened/WhenProcessingTheJob.java
package com.lasthopesoftware.bluewater.client.stored.library.items.files.job.GivenAFileThatDoesNotYetExist.AndTheInputStreamCannotBeOpened;

import com.lasthopesoftware.bluewater.client.browsing.items.media.files.ServiceFile;
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.IStoredFileAccess;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.StoredFileJob;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.StoredFileJobProcessor;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.StoredFileJobState;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.repository.StoredFile;
import com.namehillsoftware.handoff.promises.Promise;

import org.junit.BeforeClass;
import org.junit.Test;

import java.io.File;
import java.io.IOException;
import java.util.Collections;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class WhenProcessingTheJob {

	private static final StoredFile storedFile = new StoredFile(new LibraryId(6), 1, new ServiceFile(1), "test-path", true);
	private static List<StoredFileJobState> jobStates;

	@BeforeClass
	public static void before() {
		final StoredFileJobProcessor storedFileJobProcessor = new StoredFileJobProcessor(
			$ -> {
				final File file = mock(File.class);
				final File parentFile = mock(File.class);
				when(parentFile.mkdirs()).thenReturn(true);
				when(file.getParentFile()).thenReturn(parentFile);

				return file;
			},
			mock(IStoredFileAccess.class),
			(libraryId, f) -> new Promise<>(new IOException()),
			f -> false,
			f -> true,
			(is, f) -> {});

		jobStates = storedFileJobProcessor.observeStoredFileDownload(Collections.singleton(
			new StoredFileJob(new LibraryId(6), new ServiceFile(1), storedFile))).map(j -> j.storedFileJobState).toList().blockingGet();
	}

	@Test
	public void thenTheStoredFileJobIsInQueuedState() {
		assertThat(jobStates).containsExactly(
			StoredFileJobState.Queued,
			StoredFileJobState.Downloading,
			StoredFileJobState.Queued);
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/stored/library/items/files/job/GivenAFileThatDoesNotYetExist/AndTheInputStreamCannotBeOpened/WhenProcessingTheJob.kt
package com.lasthopesoftware.bluewater.client.stored.library.items.files.job.GivenAFileThatDoesNotYetExist.AndTheInputStreamCannotBeOpened

import com.lasthopesoftware.bluewater.client.browsing.files.ServiceFile
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.NullPromisingWritableStream
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.StoredFileJob
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.StoredFileJobProcessor
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.StoredFileJobState
import com.lasthopesoftware.bluewater.client.stored.library.items.files.repository.StoredFile
import com.lasthopesoftware.bluewater.shared.promises.extensions.DeferredPromise
import com.lasthopesoftware.promises.extensions.toPromise
import com.lasthopesoftware.resources.io.PromisingReadableStream
import io.mockk.every
import io.mockk.mockk
import io.reactivex.rxjava3.core.Observable
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import java.io.IOException
import java.net.URI

class WhenProcessingTheJob {
	private val storedFile = StoredFile(LibraryId(6), ServiceFile("1"), URI("test-path"), true)
	private var jobStates: List<StoredFileJobState>? = null

	@BeforeAll
	fun before() {
		val deferredDownloadPromise = DeferredPromise<PromisingReadableStream>(IOException())

		val storedFileJobProcessor = StoredFileJobProcessor(
			mockk {
				every { promiseOutputStream(any()) } returns NullPromisingWritableStream.toPromise()
			},
			mockk { every { promiseDownload(any(), any()) } returns deferredDownloadPromise },
			mockk(relaxed = true),
		)

		jobStates = storedFileJobProcessor.observeStoredFileDownload(
			Observable.just(
				StoredFileJob(LibraryId(6), ServiceFile("1"), storedFile)
			)
		).map { j ->
			j.storedFileJobState.also {
				if (it == StoredFileJobState.Downloading)
					deferredDownloadPromise.resolve()
			}
		}.toList().blockingGet()
	}

	@Test
	fun `then the stored file job is in queued state`() {
		assertThat(jobStates).containsExactly(
			StoredFileJobState.Queued,
			StoredFileJobState.Downloading,
			StoredFileJobState.Queued
		)
	}
}
